api_documentation: Render object details in parameter description.

laurynmm opened the pull request:
Adds a check for object in parameter type that will render the details of the object in the parameter description if they are in the object definition in the OpenAPI documentation.

I borrowed some of the code from `api_return_values_table_generator.py` since objects are being rendered in the documentation when they are return values. There is definitely some cleaning to do and I need to test on an object definition that contains another object, but I wanted to get feedback from folks who have worked on the api documentation markdown rendering for the changes made so far.

Fixes #19424.

**Testing plan:** [current diff of API endpoints html](https://pastebin.com/6Lw2e8ZU)

**GIFs or screenshots:** 
![Screenshot from 2021-12-01 18-14-32](https://user-images.githubusercontent.com/63245456/144286123-b58d9c80-1257-4dc3-bcd6-913448c7bdd2.png)


![Screenshot from 2021-12-01 17-44-17](https://user-images.githubusercontent.com/63245456/144286126-f2c0f10d-9860-4771-b40d-2dfebbdf597a.png)

andersk commented on zerver/lib/markdown/api_arguments_table_generator.py:
```diff
@@ -194,3 +209,33 @@ def generate_data_type(schema: Mapping[str, Any]) -> str:
+            description = object_values[value]["description"]
+            data_type = generate_data_type(object_values[value])
+            object_details += render_description(description, spacing, data_type, value)
+        object_details += "</ul>\n"
```
This `</ul>` is incorrectly omitted in the `if` branch.

timabbott commented on zerver/lib/markdown/api_arguments_table_generator.py:
```diff
@@ -194,3 +209,33 @@ def generate_data_type(schema: Mapping[str, Any]) -> str:
+        + "</span> "
+        + description
+        + "</li>\n"
+    )
```
Constructing HTML with string manipulation is error-prone, as demonstrated in the example above; I'd consider using some of the `Element` and `SubElement` classes in `zerver/lib/markdown/__init__.py` (and the upstream markdown library) for constructing HTML in the markdown processor.

Also, it can introduce security bugs were these strings user-controlled, because (for example) the content of the code block isn't being automatically HTML-escaped.

laurynmm commented:
@timabbott @andersk I've gone back through and updated my implementation for this issue, but I might still be being naive about constructing the HTML. But, as I'm still learning, I figured this would be a good point to get feedback.

I read through all of the other preprocessors in `zerver/lib/markdown` and used them (particularly `tabbed_sections.py` since it builds a list) as a guide. Most of them are much cleaner and easier to understand than the two api preprocessors.

**Testing**: I'd like to do some more manual testing in the dev environment by changing the `zerver/openapi/zulip.yaml` document around for possible objects in the definitions. I've already done a few checks for nested objects and fixed a few bugs I found. Other testing suggestions are welcome.

**Diff**: Here is the [current HTML diff](https://pastebin.com/spAhvULe) ignoring changes to white space and blank lines.

### Here's an example of my testing for nested objects in the api definition
![Screenshot from 2021-12-08 18-31-51](https://user-images.githubusercontent.com/63245456/145275218-ed3fe5c8-d96e-4ffb-86bb-275468a10ce5.png)

### Here's how the drafts object is being rendered with the changes
![Screenshot from 2021-12-08 18-20-07](https://user-images.githubusercontent.com/63245456/145275222-511602f8-df69-45e9-b0f4-f990aa8582fc.png)

timabbott commented on zerver/lib/markdown/api_arguments_table_generator.py:
```diff
@@ -163,18 +176,72 @@ def render_table(self, arguments: Sequence[Mapping[str, Any]]) -> List[str]:
+                    object_block = self.render_object_details(object_schema["items"], str(name))
+                elif "properties" in object_schema:
+                    object_block = self.render_object_details(object_schema, str(name))
+                # TODO: There are some endpoint parameters with object properties that are not defined in `zerver/openapi/zulip.yaml`
```
This seems like something we can try to eliminate in a follow-up pass; they're probably all oversights.  But I'd be happy merging this as-is, and then doing that as a next mini project; likely adding an assert in that `else` block and then running `test-backend test_docs` (I think that's the one that renders every API endpoint) should catch any entries with this issue.

laurynmm replied:
Agreed. I added the comment as a reminder thinking of that being a mini follow-up to this update. I found the error while checking the html diff on the api documentation and have the two endpoints currently without object definitions noted.

So that might be another argument for trying to figure out a way to add an html diff comparison (for both api and help documentation) as a shared tool or test for the development community. The way I have mine set up, it gives me an error log of any link that doesn't return '200' as its status code.

timabbott commented:
This looks reasonable to me!

laurynmm commented:
There are two aspects of the values of the object not currently being rendered: "example" and "required". 

The object itself has an example, so I felt like that was sufficient and having an example for each value of the object would be some duplication. Currently, the only value in the `draft` object with an example in the definition is `timestamp`.

And with "required", it read oddly when I tried to add it in the same line as the "enum" and "default" in the description text. I could go through and try to add it like it is in the top part of the parameter. But if I do, then I think it will get pulled into the `meta` descriptions in the html, if the value is required. If it's optional, then it wouldn't. For the `draft` object, the `id` and `timestamp` are listed as optional because they are not required for the endpoints that create and edit drafts. See screenshot below.

Finally, as noted in the CZO chat, the rendering for an enum that's an empty string is not ideal at the moment.

### Here's a screenshot with the addition of required or option to the end of the descriptions.
![Screenshot from 2021-12-09 18-55-50](https://user-images.githubusercontent.com/63245456/145450164-858b52c2-78a9-498c-a7a9-6ac1d0d3e637.png)

laurynmm commented:
I have managed to get the parameter object definition completely rendered in this current iteration. String enums and defaults are being rendered in code blocks with quotations. Any information, like an example or default or required / optional, is being rendered with the value's description. [Current html diff](https://pastebin.com/spAhvULe) when compared to the main branch.

I would appreciate a look through the code. I feel like the way I managed to get the default value of a string enum to render was a bit messy.

**Screenshot**: shows examples of different defaults, enums, required / optional settings and if there is an example for a value

![Screenshot from 2021-12-10 15-50-39](https://user-images.githubusercontent.com/63245456/145601013-ab7e9bbf-7da7-460f-a0f1-09c35524f623.png)

laurynmm commented:
@timabbott I saw that you've marked this as a release goal! I wanted to check-in about reviewing my updated commit so that I can get some feedback about how to proceed. In the meantime, I've been prepping for some of the follow-up tasks I mentioned in my last comment. Thanks!

laurynmm commented:
Updates as of Jan 17, 2022:
- [Updated HMTL diff](https://pastebin.com/QenuYP11) for latest updates. Still shows that the only effected doc endpoints are `api/create-drafts` and `api/edit-draft`.
- I realized the previous version was not handling nested lists correctly. I think the current version does the recursive call in the correct place to fix that issue. Manually tested in the dev env with some fake additions to the OpenAPI (see GIF below).
- I removed the empty paragraph that was generated when there was no additional information (default values, enum values, required or optional value, example of value) to the description text. And removed the white space that was generated when only some of those values were present.

**GIF of nested lists**:
![Peek 2022-01-17 19-07](https://user-images.githubusercontent.com/63245456/149819827-9b747e33-7061-423f-bb80-7e332091aaa5.gif)

timabbott commented:
Merged, thanks @laurynmm!

I think we may want to revisit the exact formatting -- e.g. I would prefer the optional/required indications match the formatting we use for optional/required parameters in requests, rather than being a separate format.

But I think it's better to do that refinement on top of something that's addressed the recursion issue properly, as this has.

laurynmm commented:
@timabbott - Yes, I agree that matching the formatting (and probably placement too) of optional/required is another follow-up to this change. And having that information available in the documentation structure vs only the written descriptions is definitely a good middle step to land on for now.
